File: 409.longest-palindrome.java
/*
 * @lc app=leetcode id=409 lang=java
 *
 * [409] Longest Palindrome
 */

// 95/95 cases passed (1 ms)
// Your runtime beats 100 % of java submissions
// Your memory usage beats 6.67 % of java submissions (37.6 MB)

// @lc code=start
class Solution {
    public int longestPalindrome(String s) {
        int[] cnts = new int[256];
        int res = 0;
        for(char c:s.toCharArray()) {
            cnts[c]++;
        }
        for(int cnt:cnts) {
            res+=(cnt/2)*2;
        }
        if(res<s.length()) {
            res++;
        }
        return res;
    }
}
// @lc code=end

File: 409.longest-palindrome.js
/*
 * @lc app=leetcode id=409 lang=javascript
 *
 * [409] Longest Palindrome
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var longestPalindrome = function(s) {
    let map = {}
    let res = 0
    for(let i=0, l=s.length; i<l; i++) {
        let ch = s[i]
        if(!map[ch]) {
            map[ch] = 1
        } else {
            map[ch] ++
        }
    }
    for(key in map) {
        if(map[key] > 1) {
            res += Math.floor(map[key]/2)
        }
    }
    res = res*2
    if(res < s.length)
        res ++

    return res
};
// longestPalindrome('abccccdd')
// @lc code=end

